# HG changeset patch
# User Punarbasu Purkayastha <ppurka@gmail.com>
# Date 1333432924 -28800
# Node ID 6fe2ada9c586aa6b548050094ec02c00619447bc
# Parent  aa148c3c0d7b06fd049b9e845eebeda2421d32bf
fix extraneous points at z=0 in list_plot3d and allow color and rgbcolor as keywords

diff --git a/sage/plot/plot3d/list_plot3d.py b/sage/plot/plot3d/list_plot3d.py
--- a/sage/plot/plot3d/list_plot3d.py
+++ b/sage/plot/plot3d/list_plot3d.py
@@ -23,9 +23,27 @@
       - a list of lists (all of the same length) - this is treated the same as
         a matrix.
     
-    - ``texture`` - (default: "automatic", a solid light blue)
+    - ``texture`` - (default: "automatic", a solid light blue) this can be
+
+      - a named color, ex. 'blue', 'black', 'red'
+
+      - an RGB 3-tuple of numbers between 0 and 1, e.g. (0.5, 0.5, 0)
+
+      - a string representing a html color, e.g. '#ffffff', '#00f'
+
+    Note that only one of ``texture``, ``color`` or ``rgbcolor`` should be
+    used. In case two or more of these options are present, the order of
+    preference is ``texture``, ``color``, ``rgbcolor``. This means that
+    ``texture`` will override ``color`` if both the options are provided,
+    and so on.
     
     OPTIONAL KEYWORDS:
+
+    - ``color`` - This is the same as ``texture``. Use only one of
+      ``texture`` or ``color`` or ``rgbcolor``.
+
+    - ``rgbcolor`` - This is the same as ``texture``. Use only one of
+      ``texture`` or ``color`` or ``rgbcolor``.
     
     - ``interpolation_type`` - 'linear', 'nn' (nearest neighbor), 'spline'
     
@@ -159,10 +177,32 @@
         Traceback (most recent call last):
         ... 
         ValueError: We need at least 3 points to perform the interpolation
+
+    Order of preference of color options is texture, color, rgbcolor. So,
+    the first plot below is colored red, while the second one is colored
+    black.::
+
+        sage: list_plot3d([(0,0,1), (2,3,4)], texture='red', color='black')
+        sage: list_plot3d([(0,0,1), (2,3,4)], color='black', rgbcolor='#0f0')
     """
     import numpy
+
     if texture == "automatic":
-        texture = "lightblue"
+        if 'color' in kwds:
+            texture = kwds.pop('color')
+        elif 'rgbcolor' in kwds:
+            texture = kwds.pop('rgbcolor')
+        else:
+            texture = "lightblue"
+    # Delete the color options that are not needed any more. texture has
+    # the correct value.
+    for s in ('color', 'rgbcolor'):
+        if s in kwds:
+            del kwds[s]
+
+    from sage.plot.colors import rgbcolor
+    texture = rgbcolor(texture)
+
     if is_Matrix(v):
         if interpolation_type=='default' or interpolation_type=='linear' and not kwds.has_key('num_points'):
             return list_plot3d_matrix(v, texture=texture,  **kwds)
@@ -411,7 +451,6 @@
 
         T= delaunay.Triangulation(x,y)
         f=T.linear_interpolator(z)
-        f.default_value=0.0
         j=numpy.complex(0,1)
         vals=f[ymin:ymax:j*num_points,xmin:xmax:j*num_points]
         from parametric_surface import ParametricSurface
@@ -432,7 +471,6 @@
 
         T=delaunay.Triangulation(x,y)
         f=T.nn_interpolator(z)
-        f.default_value=0.0
         j=numpy.complex(0,1)
         vals=f[ymin:ymax:j*num_points,xmin:xmax:j*num_points]
         from parametric_surface import ParametricSurface
